feature(walledgarden): routing now respects walled garden policies

hypeJunction opened the pull request:
Adds a new Router method to decouple routing logic from ElggSite

BREAKING CHANGE
Walled garden initialization has been decoupled from the site instance,
and is now part of the routing policies.
ElggSite::checkWalledGarden() and ElggSite::isPublicPage() have been removed.

Refs #9881

hypeJunction commented:
@mrclay First stab.

hypeJunction commented on engine/classes/Elgg/Router.php:
```diff
@@ -184,5 +201,74 @@ public function allowRewrite(Request $request) {
+	 * @return bool
+	 * @since 2.2
+	 */
+	public function isPublicPage($url = '', $site_guid = 0) {
```
We could inject `\Elgg\Http\Request` and `ElggSite` here.

mrclay commented on engine/classes/Elgg/Router.php:
```diff
@@ -184,5 +201,74 @@ public function allowRewrite(Request $request) {
+
+		// allow public pages
+		foreach (array_merge($defaults, $plugins) as $public) {
+			$pattern = "`^{$site_url}{$public}/*$`i";
```
We should use preg_quote() on $site_url.

mrclay commented:
I'm fine with the direction. My main issue is the branch; this no longer protects unrouted pages (that include `start.php`) and this isn't something I'd expect to change in a minor release.

hypeJunction commented:
How does this sound:
In 2.x
- deprecate the APIs and proxy them to Router.
- move hook registrations into init handler
- move forward() to the init function and add the `ELGG_BYPASS_WALLED_GARDEN` flag

In 3.0
- add the breaking change and only perform public page check in the Router. By that time we should already have the new HTTP API, so it should be easier to deal with it.

mrclay commented:
With everything already in 2.2, I'd feel safer just adding the simple-to-implement bypass switch. There's not a huge cry out for #9881 and we can work these improvements into 2.3 and 3. What do you think?

hypeJunction commented:
Let's leave it for now. We can work on it while working on HTTP responses.

hypeJunction commented:
This is BC now
Made some changes to how we handle HTTP responses
Added tests for walled garden

mrclay commented on engine/classes/Elgg/Http/ResponseFactory.php:
```diff
@@ -144,7 +144,10 @@ public function send(Response $response) {
 			}
 		} else {
-			$response->send();
+			if (elgg_trigger_before_event('send', 'http_response', $response)) {
```
If cancelled, what should happen? Certainly the after event shouldn't be called.

mrclay replied:
Another case for event params. As is handlers don't receive any contextual info that would help them to make a decision here.

hypeJunction replied:
My assumption is that the response is always sent, the question is how - using the default method, or i.e. writing to log file.

hypeJunction replied:
Aside from PHPUnit applicability, my use case is here: https://github.com/hypeJunction/elgg-cli/blob/master/src/Elgg/CLI/Command.php#L38.
I envision, this will come in handy when working with web services - ultimately, we could allow calling web services by route and having control over how the data is returned will be important.

mrclay commented on engine/classes/Elgg/Router.php:
```diff
@@ -193,4 +209,74 @@ public function allowRewrite(Request $request) {
+	 * @param int    $site_guid Optional. GUID of the site the URL belongs to.
+	 *
+	 * @return bool
+	 * @since 2.2
```
2.3, sorry.

mrclay commented on docs/guides/events-list.rst:
```diff
@@ -65,6 +65,12 @@ System events
 **cache:flush, system**
     Reset internal and external caches, by default including system_cache, simplecache, and memcache. One might use it to reset others such as APC, OPCache, or WinCache.
 
+**send:before, http_response**
```
Could this go in a separate PR? Would be nice to assemble use cases and cancellation behavior.

hypeJunction replied:
Can cherry pick into another PR: https://github.com/Elgg/Elgg/pull/9883/commits/e19846a6f62ba3cd79c156d51e0b2115f8639e2f

hypeJunction replied:
Alternative approach: #10064

hypeJunction commented on engine/tests/phpunit/Elgg/TestCase.php:
```diff
@@ -4,10 +4,16 @@
 
-abstract class TestCase extends \PHPUnit_Framework_TestCase {
+/**
+ * @runTestsInSeparateProcesses
```
You would think that this would slow down tests, but it made them faster. I think because our services are quite massive preserving global states in fact makes things worse.

hypeJunction commented:
Rebased on top of #10064

hypeJunction commented:
Reopening now that walled garden is getting some love from @jdalsem

hypeJunction commented:
Rebased for master

jdalsem commented:
Could you fix https://github.com/Elgg/Elgg/issues/7235 while at it?

hypeJunction commented:
Rebased
Fixed #7235

hypeJunction committed:
feat(walledgarden): router now respects walled garden policies

Walled garden routing has been decoupled from the ElggSite instance
and moved into the Router.
"public_pages,walled_garden" hook now receives system public routes
as the default value.

Fixes #7235
Fixes #9881

BREAKING CHANGE
ElggSite is no longer responsible for enforcing walled garden policies:
this logic has been moved to the Router.
ElggSite::checkWalledGarden() and ElggSite::isPublicPage() have been
removed. There is no longer public API to test a URL against walled garden
policies - plugins should route their pages properly through the router.
